torch/csrc/jit/passes/mkldnn_rewrite.cpp

```diff
@@ -37,32 +37,45 @@ c10::VaryingShape<int64_t> getSizesOf(Node* n, size_t idx) {
   return tt->sizes();
 }
 
-bool ndimEquals(c10::VaryingShape<int64_t> sizes, size_t value) {
-  if (!sizes.concrete_sizes()) {
+bool isContiguous(Value* v) {
+  auto const& tt = v->type()->cast<TensorType>();
+  if (!tt) {
     return false;
   }
-  auto concrete_sizes = *sizes.concrete_sizes();
-  if (concrete_sizes.size() != value) {
+  if (!tt->isComplete()) {
     return false;
   }
-  return true;
+  auto const& sizes = tt->sizes().concrete_sizes();
+  auto const& strides = tt->strides().concrete_sizes();
+  if (!sizes || !strides) {
+    return false;
+  }
+  int ndims = (*sizes).size();
+  // TODO: only support 4D input for now
+  if (ndims != 4) {
+    return false;
+  }
+
+  // TODO: currently only support contiguous input. Channels last contiguous
+  // will be supported as a next step
+  return *strides == TensorType::contiguousStridesOf(*sizes);
```

Comment:
Do we have any difference in graph rewrite between CL cont and legacy cont?

Reply:
TE does not support CL cont for now.
If we write the graph to our prepack OP when the input is CL cont, since we've enabled TE to select our prepack OP into the TE fusion group, our prepack OP will run as legacy cont (TE fusion group will always convert the input to be legacy cont), which may bring performance regression compared with the one without this PR.

### If we rewrite CL cont in this PR:
```
with prim::TensorExprGroup_0 = graph(%x.1 : Float(1, 3, 224, 224, strides=[150528, 1, 672, 3], requires_grad=0, device=cpu)):
  %1 : __torch__.torch.classes.mkldnn.ConvOpContext = prim::Constant[value=object(0x560fa1a0d2c0)]()
  %2 : Float(1, 10, 222, 222, strides=[492840, 1, 2220, 10], requires_grad=0, device=cpu) = mkldnn_prepacked::conv2d_run(%x.1, %1)
  return (%2)
```
1. TensorExprGroup will convert CL input to be NCHW cont.
2. Run into our `mkldnn_prepacked::conv2d_run` kernel with input format `abcd`:
```
onednn_verbose,exec,cpu,convolution,jit:avx512_core,forward_training,src_f32::blocked:abcd:f0 wei_f32:p:blocked:Acdb16a:f0 bia_undef::undef::f0 dst_f32:p:blocked:aBcd16b:f0,,alg:convolution_direct,mb1_ic3oc10_ih224oh222kh3sh1dh0ph0_iw224ow222kw3sw1dw0pw0,0.775146
onednn_verbose,exec,cpu,reorder,jit:uni,undef,src_f32:p:blocked:aBcd16b:f0 dst_f32::blocked:abcd:f0,attr-scratchpad:user ,,1x10x222x222,0.466064
```
3. TensorExprGroup will convert NCHW cont output back to CL cont.

### Current master behavior without this PR
`aten::conv2d` won't be selected into TE fusion group thus it runs with input in `acdb` directly:
```
onednn_verbose,exec,cpu,reorder,jit:uni,undef,src_f32::blocked:acdb:f0 dst_f32:p:blocked:Acdb16a:f0,attr-scratchpad:user ,,10x3x3x3,0.00195312
onednn_verbose,exec,cpu,convolution,brgconv:avx512_core,forward_training,src_f32::blocked:acdb:f0 wei_f32:p:blocked:Acdb16a:f0 bia_undef::undef::f0 dst_f32::blocked:acdb:f0,attr-scratchpad:user ,alg:convolution_direct,mb1_ic3oc10_ih224oh222kh3sh1dh0ph0_iw224ow222kw3sw1dw0pw0,0.253906
```

We will support CL cont once TE supports CL cont.